### PR TITLE
[SYCL] Add Unified Runtime plugin and route to it with SYCL_PREFER_UR

### PR DESCRIPTION
**Pull request by smaslov-intel**

### CODE REVIEW

File: sycl/source/detail/platform_impl.cpp [hunk 1]
```diff
@@ -140,6 +134,60 @@ std::vector<platform> platform_impl::get_platforms() {
...
+    return Platforms;
+  };
+
+  static const bool PreferUR = [] {
+    const char *PreferURStr = std::getenv("SYCL_PREFER_UR");
+    return (PreferURStr && (std::stoi(PreferURStr) != 0));
+  }();
+
+  // See which platform we want to be served by which plugin.
+  // There should be just one plugin servins each backend.
```
Comment (jandres742) [inline sycl/source/detail/platform_impl.cpp:146]: serving

**Review by jandres742 — APPROVED** [review]

File: sycl/unittests/pi/EnqueueMemTest.cpp [hunk 2]
```diff
@@ -109,7 +109,7 @@ INSTANTIATE_TEST_SUITE_P(
     EnqueueMemTestImpl, EnqueueMemTest,
     testing::ValuesIn(pi::initializeAndRemoveInvalid()),
     [](const testing::TestParamInfo<EnqueueMemTest::ParamType> &info) {
-      return pi::GetBackendString(info.param);
+      return pi::GetBackendString(info.param).c_str();
```
Comment (againull) [inline sycl/unittests/pi/EnqueueMemTest.cpp:112]: Unfortunately, this also causes "error: returning address of local temporary object"

File: sycl/unittests/pi/PlatformTest.cpp [hunk 3]
```diff
@@ -66,7 +66,7 @@ INSTANTIATE_TEST_SUITE_P(
     PlatformTestImpl, PlatformTest,
     testing::ValuesIn(pi::initializeAndRemoveInvalid()),
     [](const testing::TestParamInfo<PlatformTest::ParamType> &info) {
-      return pi::GetBackendString(info.param);
+      return pi::GetBackendString(info.param).c_str();
```
Comment (againull) [inline sycl/unittests/pi/PlatformTest.cpp:69]: Unfortunately, this also causes "error: returning address of local temporary object"

Comment (againull) [inline sycl/unittests/pi/PlatformTest.cpp:69]:
```suggestion
      return pi::GetBackendString(info.param);
```

File: sycl/unittests/pi/EnqueueMemTest.cpp (hunk 2 above)
Comment (againull) [inline sycl/unittests/pi/EnqueueMemTest.cpp:112]:
```suggestion
      return pi::GetBackendString(info.param);
```

File: sycl/unittests/pi/PlatformTest.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/unittests/pi/PlatformTest.cpp:69]: indeed, returning "std::string" now

**Comment by smaslov-intel** [comment]
@againull , @cperkinsintel , @steffenlarsen : please review this

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by smaslov-intel** [comment]
The testing found subtle issue that there were multiple "plugin" objects representing the same backend floating around that got out-of-sync. To make the whole thing more robust I needed to wrap all uses of "plugin" with shared_ptr: https://github.com/intel/llvm/pull/9232/commits/f23631761456030edc1d3e7ae5ec6d05fd561c63

**Comment by smaslov-intel** [comment]
> The testing found subtle issue that there were multiple "plugin" objects representing the same backend floating around that got out-of-sync. To make the whole thing more robust I needed to wrap all uses of "plugin" with shared_ptr: [f236317](https://github.com/intel/llvm/commit/f23631761456030edc1d3e7ae5ec6d05fd561c63)

@intel/llvm-gatekeepers : I think this is finally good to get merged. Please review. This failure is handled in https://github.com/intel/llvm/commit/d56e0ae25f859aa76277b6d5882203c77ed3195a:
>  SYCL :: Scheduler/DataMovement.cpp

**Comment by steffenlarsen** [comment]
No applicable L0 reviewers available. Trusting @jandres742 for the small related L0 backend changes.

**Pull request merged**